src/Domain/TagsCharacteresticSpeeds.hpp

```diff
@@ -0,0 +1,58 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <boost/optional.hpp>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"  // For Tags::Normalized
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Tags.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace domain {
+namespace Tags {
+/// Compute the characteristic speeds on the moving mesh given the
+/// characteristic speeds if the mesh were stationary.
+///
+/// \note Assumes that `typename CharSpeedsComputeTag::return_type` is a
+/// `std::array<DataVector, NumberOfCharSpeeds>`
+template <typename CharSpeedsComputeTag, size_t Dim>
+struct CharSpeedCompute : CharSpeedsComputeTag {
+  using base = CharSpeedsComputeTag;
+  using return_type = typename CharSpeedsComputeTag::return_type;
+
+  template <typename... Ts, typename T, size_t NumberOfCharSpeeds>
+  static void function(
+      const gsl::not_null<std::array<T, NumberOfCharSpeeds>*> result,
+      const boost::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+          grid_velocity,
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& unit_normal_covector,
+      const Ts&... ts) noexcept {
+    // Note that while the CharSpeedsComputeTag almost certainly also needs the
+    // unit normal covector for computing the original characteristic speeds, we
+    // don't know which of the `ts` it is, and thus we need the unit normal
+    // covector to be passed explicitly.
+    CharSpeedsComputeTag::function(result, ts...);
+    if (static_cast<bool>(grid_velocity)) {
+      const Scalar<DataVector> normal_dot_velocity =
+          dot_product(*grid_velocity, unit_normal_covector);
+      for (size_t i = 0; i < result->size(); ++i) {
+        gsl::at(*result, i) -= get(normal_dot_velocity);
+      }
+    }
+  }
+
+  using argument_tags =
+      tmpl::append<tmpl::list<MeshVelocity<Dim, Frame::Inertial>,
```

Comment:
use `push_front`

Reply:
For my education: why this change request? Is it a compiler optimization where `push_front<...>` is more efficient than `append<list<...>>`? Or something else I am missing?

Reply:
`append` will do the wrong thing if you `append<list<...>, item>`, you will get an empty list,
you can also push multiple items to the back or front of the list

Reply:
Ah, excellent question! So `tmpl::append<Tags::A, some_list>` silently gives you an empty list (which is annoying and leads to bugs) so we prefer `tmpl::push_back` and `tmpl::push_front` since they don't have this weird behavior.

Reply:
Ah yes, I'd forgotten about this bug. Thank you both for the quick responses :)